pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -242,90 +317,276 @@ func (sf *IteratorDatastore) Read(
-		// a clone from it.
-	} else {
-		if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
-			sf.internalStorage.mu.Unlock()
-			sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationRead).Inc()
-
-			// we cannot share this iterator because we have reached the size limit.
+		item, err := keyItem.iter.clone()
+		if err != nil {
+			keyItem.iter.mu.Unlock()
+			sf.deref(cacheKey)
+			telemetry.TraceError(span, err)
 			return sf.RelationshipTupleReader.Read(ctx, store, tupleKey, options)
 		}
-
-		sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
-			counter: 1,
-		}
+		keyItem.iter.mu.Unlock()
+		sharedIteratorQueryHistogram.WithLabelValues(
+			storagewrappersutil.OperationRead, "true",
+		).Observe(float64(time.Since(start).Milliseconds()))
+		return item, nil
+	}
+	if len(sf.internalStorage.iters) >= sf.internalStorage.limit {
 		sf.internalStorage.mu.Unlock()
-		span.SetAttributes(attribute.Bool("found", false))
+		sharedIteratorBypassed.WithLabelValues(storagewrappersutil.OperationRead).Inc()
+
+		// we cannot share this iterator because we have reached the size limit.
+		return sf.RelationshipTupleReader.Read(ctx, store, tupleKey, options)
 	}
+
+	newIterator := newSharedIterator(sf, cacheKey, sf.watchdogTimeoutConfig)
+	newIterator.mu.Lock()
+
+	sf.internalStorage.iters[cacheKey] = &internalSharedIterator{
+		counter: 1,
+		iter:    newIterator,
+	}
+	sf.internalStorage.mu.Unlock()
+	span.SetAttributes(attribute.Bool("found", false))
+
 	actual, err := sf.RelationshipTupleReader.Read(ctx, store, tupleKey, options)
 	if err != nil {
-		_, _ = sf.deref(cacheKey)
+		newIterator.queryErr = err
+		newIterator.mu.Unlock()
+
+		sf.deref(cacheKey)
+		telemetry.TraceError(span, err)
 		return nil, err
 	}
+	newIterator.inner = actual
+	newIterator.mu.Unlock()
+
 	sharedIteratorQueryHistogram.WithLabelValues(
-		storagewrappersutil.OperationRead, strconv.FormatBool(found),
+		storagewrappersutil.OperationRead, "false",
 	).Observe(float64(time.Since(start).Milliseconds()))
-
-	return newSharedIterator(sf, cacheKey, actual), nil
+	return newIterator, nil
 }
 
+// cleanup the key and associated
+
 // decrement cacheKey from internal reference count
 // If reference count is 0, remove cacheKey.
 // Return whether item is removed.
-func (sf *IteratorDatastore) deref(cacheKey string) (bool, error) {
+func (sf *IteratorDatastore) deref(cacheKey string) {
 	sf.internalStorage.mu.Lock()
-	defer sf.internalStorage.mu.Unlock()
 	item, ok := sf.internalStorage.iters[cacheKey]
 	if !ok {
+		sf.internalStorage.mu.Unlock()
+
 		sf.logger.Error("failed to dereference cache key", zap.String("cacheKey", cacheKey))
-		return false, fmt.Errorf("failed to dereference cache key: %s", cacheKey)
+		return
 	}
 	item.counter--
 	if item.counter == 0 {
 		delete(sf.internalStorage.iters, cacheKey)
-		return true, nil
+		sf.internalStorage.mu.Unlock()
+		item.iter.cleanup()
```

Comment:
To clean this up you can queue up `defer` statements so guaranteeing that this operation happens after the yield of `sf.internalStorage.mu`.

Reply:
As noted above, the defer is first in last out.  As such, we cannot use defer because we need to unlock before cleanup.